jina/drivers/multimodal.py

```diff
@@ -0,0 +1,114 @@
+                |
+                |- chunk: {modality: mode2}
+        Output:
+        document: (embedding: multimodal encoding)
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+
+    .. note::
+        - It traverses on the ``documents`` for which we want to apply the ``multimodal`` embedding. This way
+        we can use the `batching` capabilities for the `executor`.
+    .. warning::
+        - It assumes that every ``chunk`` of a ``document`` belongs to a different modality.
+    """
+
+    def __init__(self,
+                 traversal_paths: Tuple[str] = ('r',), *args, **kwargs):
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
+
+    @property
+    def position_by_modality(self):
+        if not getattr(self._exec, 'position_by_modality', None):
+            raise RuntimeError('Could not know which position of the ndarray to load to each modality')
+        return self._exec.position_by_modality
+
+    def _get_executor_input_arguments(self, content_by_modality: Dict[str, 'np.ndarray'], num_modalities: int):
+        """
+            From a dictionary ``content_by_modality`` it returns the arguments in the proper order so that they can be
+            passed to the executor.
+        """
+        input_args = [None] * num_modalities
+        for modality in self.position_by_modality.keys():
+            input_args[self.position_by_modality[modality]] = content_by_modality[modality]
+        return input_args
+
+    def _apply_all(
+            self,
+            docs: Iterable['jina_pb2.Document'],
+            *args, **kwargs
+    ) -> None:
+        """
+        :param docs: the docs for which a ``multimodal embedding`` will be computed, whose chunks are of different
+        modalities
+        :return:
+        """
+        num_modalities = len(self.position_by_modality.keys())
+        content_by_modality = defaultdict(list)  # array of num_rows equal to num_docs and num_columns equal to
+
+        valid_docs = []
+        for doc in docs:
+            doc_content = _extract_modalities_from_document(doc)
+            if doc_content:
+                valid_docs.append(doc)
+                for modality in self.position_by_modality.keys():
+                    content_by_modality[modality].append(doc_content[modality])
+            else:
+                self.logger.warning(f'Invalid doc {doc.id}. Only one chunk per modality is accepted')
+
+        if len(valid_docs) > 0:
+            # I want to pass a variable length argument (one argument per array)
+            for modality in self.position_by_modality.keys():
+                content_by_modality[modality] = np.stack(content_by_modality[modality])
+
+            # Guarantee that the arguments are provided to the executor in its desired order
+            input_args = self._get_executor_input_arguments(content_by_modality, num_modalities)
+            embeds = self.exec_fn(*input_args)
```

Comment:
to me this part looks like a fancy `concat`, we do have a concat driver, not sure how they are related.

Reply:
It is something that would be able to replace concatenating actually. You can see ConcatDriver as the combination of a `ReduceAll`plus a `MultiModalDriver` equipped with a `ConcatExecutor`

Reply:
Concatenating is just another way of combining different modalities into one single vectorial representation. In the tests we have tested with such executors for instance